File: dataset_generator.py
import random
import string

def generate_random_string(length): 
    # Generate a random string of specified length using letters (both cases)
    letters = string.ascii_letters
    return ''.join(random.choice(letters) for _ in range(length))

def generate_dataset(settings, output_file):
    num_to_generate = int(input("Enter the number of records to generate: "))
    with open(output_file, 'w') as f:
        for _ in range(num_to_generate):
            # Generate random number between 1 and 1 million
            number = random.randint(settings["start_number"], settings["end_number"])
            
            # Randomly choose between 4 or 5 characters
            str_length = random.randint(settings["string_length_min"], settings["string_length_max"])
            random_string = generate_random_string(str_length)
            
            # Write to file in the format: number,string
            f.write(f"{number},{random_string}\n")

def setting_dataset_type():
    start_number = int(input("Enter the start number: "))
    end_number = int(input("Enter the end number: "))
    string_length_min = int(input("Enter the minimum length of the string: "))
    string_length_max = int(input("Enter the maximum length of the string: "))
    return {"start_number":start_number,"end_number":end_number,"string_length_min":string_length_min,"string_length_max":string_length_max}

default = {"start_number":1,"end_number":1000000000,"string_length_min":4,"string_length_max":5}

if __name__ == "__main__":
    generate_type = input("Enter the type of dataset to generate: "
                            "\n1. default (assignment)"
                            "\n2. settings"
                            "\nInput:")
    if generate_type == "1":
        pass
    elif generate_type == "2":
        default = setting_dataset_type()
    generate_dataset(default,"dataset_1000000.csv")

    print("Dataset generated successfully in dataset_1000000.csv")

File: dataset_generator.cpp
#include <iostream>
#include <fstream>
#include <string>
#include <random>
#include <chrono>

using namespace std;

// Structure to hold dataset settings
struct Settings {
    int start_number;
    int end_number;
    int string_length_min;
    int string_length_max;
};

// Function to generate random string, given min max of string length
string generate_random_string(int length_min, int length_max) {
    const string letters = "abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ";
    random_device seed; //Creates a seed
    mt19937 gen(seed()); //Creates a random number generator
    uniform_int_distribution<> dist(0, letters.size() - 1); //So that it will choose a number between 0 and the size of letters
    
    string result;
    for(int i = 0; i < length_min; ++i) {
        result += letters[dist(gen)];
    }
    for (int i = length_min; i < length_max; ++i) {
        if (rand()%2 == 0) {
            result += letters[dist(gen)];
        }
    }
    return result;
}

// Function to generate random number, given min max of number
string generate_random_number(int start, int end) {
    random_device seed;
    mt19937 gen(seed());
    uniform_int_distribution<> dist(start, end);
    return to_string(dist(gen));
}

// Function to generate dataset
void generate_dataset(const Settings& settings, const string& output_file) {
    int num_to_generate;
    cout << "Enter the number of records to generate: ";
    cin >> num_to_generate;

    ofstream file(output_file);
    if (!file.is_open()) {
        cerr << "Error: Could not open file " << output_file << endl;
        return;
    }

    for(int i = 0; i < num_to_generate; ++i) {
        string random_number = generate_random_number(settings.start_number, settings.end_number);
        string random_string = generate_random_string(settings.string_length_min, settings.string_length_max);
        
        file << random_number << "," << random_string << "\n";
    }

    file.close();
}

// Function to set custom settings
Settings setting_dataset_type() {
    Settings settings;
    cout << "Enter the start number: ";
    cin >> settings.start_number;
    cout << "Enter the end number: ";
    cin >> settings.end_number;
    cout << "Enter the minimum length of the string: ";
    cin >> settings.string_length_min;
    cout << "Enter the maximum length of the string: ";
    cin >> settings.string_length_max;
    return settings;
}

int main() {
    // Default settings
    Settings default_settings = {1, 1000000000, 4, 5}; //1 to 1000000000, 4 to 5 characters
    
    cout << "Enter the type of dataset to generate:\n"
         << "1. default (assignment)\n"
         << "2. settings\n"
         << "Input: ";
    
    string generate_type;
    cin >> generate_type;

    Settings settings = default_settings;
    if (generate_type == "2") {
        settings = setting_dataset_type();
    }
    //if 1, generate a dataset with default settings
    generate_dataset(settings, "dataset_1000000.csv");
    cout << "Dataset generated successfully in dataset_1000000.csv" << endl;

    return 0;
} 